lib/Dialect/FIRRTL/Transforms/DropConst.cpp

```diff
@@ -0,0 +1,128 @@
+                  element.type = getAllConstDroppedType(element.type);
+                  return element;
+                }));
+        return BundleType::get(type.getContext(), constDroppedElements, false);
+      })
+      .Case<FVectorType>([](FVectorType type) {
+        return FVectorType::get(getAllConstDroppedType(type.getElementType()),
+                                type.getNumElements(), false);
+      })
+      .Case<FEnumType>([](FEnumType type) {
+        SmallVector<FEnumType::EnumElement> constDroppedElements(
+            llvm::map_range(
+                type.getElements(), [](FEnumType::EnumElement element) {
+                  element.type = getAllConstDroppedType(element.type);
+                  return element;
+                }));
+        return FEnumType::get(type.getContext(), constDroppedElements, false);
+        ;
+      })
+      .Default([](Type) {
+        llvm_unreachable("unknown FIRRTL type");
+        return FIRRTLBaseType();
+      });
+}
+// NOLINTEND(misc-no-recursion)
+
+/// Returns null type if no conversion is needed.
+static FIRRTLBaseType convertType(FIRRTLBaseType type) {
+  auto nonConstType = getAllConstDroppedType(type);
+  return nonConstType != type ? nonConstType : FIRRTLBaseType{};
+}
+
+/// Returns null type if no conversion is needed.
+static Type convertType(Type type) {
+  if (auto base = type.dyn_cast<FIRRTLBaseType>()) {
+    return convertType(base);
+  }
+
+  if (auto refType = type.dyn_cast<RefType>()) {
+    if (auto converted = convertType(refType.getType()))
+      return RefType::get(converted, refType.getForceable());
+  }
+
+  return {};
+}
+
+namespace {
+class DropConstPass : public DropConstBase<DropConstPass> {
+  void runOnOperation() override {
+    auto module = getOperation();
+    auto fmodule = dyn_cast<FModuleOp>(*module);
+
+    // Convert the module body if present
+    if (fmodule) {
+      fmodule->walk([](Operation *op) {
+        if (auto constCastOp = dyn_cast<ConstCastOp>(op)) {
+          // Remove any `ConstCastOp`, replacing results with inputs
+          constCastOp.getResult().replaceAllUsesWith(constCastOp.getInput());
+          constCastOp->erase();
+          return;
+        }
+
+        for (auto result : op->getResults()) {
+          if (auto convertedType = convertType(result.getType()))
+            result.setType(convertedType);
+        }
+      });
+    }
+
+    // Find 'const' ports
+    auto portTypes = SmallVector<Attribute>(module.getPortTypes());
+    for (size_t portIndex = 0, numPorts = module.getPortTypes().size();
+         portIndex < numPorts; ++portIndex) {
+      if (auto convertedType = convertType(module.getPortType(portIndex))) {
+        // If this is an FModuleOp, register the block argument to drop 'const'
+        if (fmodule)
+          fmodule.getArgument(portIndex).setType(convertedType);
+        portTypes[portIndex] = TypeAttr::get(convertedType);
+      }
+    }
```

Comment:
Update `module.getPortTypes().size()` -> `portTypes.size()`

If you don't need the index anymore (because the block argument type has already been updated).
```c++
  SmallVector<Attribute> portTypes;
  portTypes.reserve(module.getNumPorts());
  llvm::transform(module.getTypes(), std::back_inserter(portTypes), [&](Attribute type) {
    if (auto convertedType = convertType(type))
      return converted type;
    return type;
  }));
```

Reply:
Updated